components/camera/rtsp/rtsp_test.go

```diff
@@ -0,0 +1,67 @@
+package rtsp
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/bhaney/rtsp-simple-server/server"
+	"github.com/edaniels/golog"
+	ffmpeg "github.com/u2takey/ffmpeg-go"
+	"go.viam.com/test"
+	viamutils "go.viam.com/utils"
+	"go.viam.com/utils/artifact"
+)
+
+func startFFMPEG(outputURL string, streamStarted, streamStopped chan struct{}) context.CancelFunc {
+	// run the test avi in a loop through ffmpeg and hand it to the server
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	viamutils.PanicCapturingGo(func() {
+		testMP4 := artifact.MustPath("components/camera/rtsp/earth_480_mjpeg.avi")
+		ffmpegStream := ffmpeg.Input(testMP4, ffmpeg.KwArgs{"re": "", "stream_loop": -1})
+		ffmpegStream = ffmpegStream.Output(outputURL, ffmpeg.KwArgs{"vcodec": "mjpeg", "huffman": "0", "f": "rtsp", "rtsp_transport": "tcp"})
+		ffmpegStream.Context = cancelCtx
+		cmd := ffmpegStream.OverWriteOutput().Compile()
+		close(streamStarted)
+		if err := cmd.Run(); err != nil {
+			close(streamStopped)
+			return
+		}
+	})
+	return cancel
+}
+
+func TestRTSPCamera(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+	// set up the rtsp simple server running on rtsp://localhost:8598/mystream
+	outputURL := "rtsp://127.0.0.1:8598/mystream"
+	configLoc := []string{artifact.MustPath("components/camera/rtsp/rtsp-simple-server.yml")}
+	s, ok := server.New(configLoc)
+	test.That(t, ok, test.ShouldBeTrue)
+	defer s.Close()
+	streamStarted := make(chan struct{})
+	streamStopped := make(chan struct{})
+	cancel := startFFMPEG(outputURL, streamStarted, streamStopped)
+	defer cancel()
+	// create the rtsp camera model
+	rtspConf := &Attrs{Address: outputURL}
+	<-streamStarted
+	rtspCam, err := NewRTSPCamera(context.Background(), rtspConf, logger)
+	// keep trying until RTSP server is running
+	for err != nil && (strings.Contains(err.Error(), "404") || strings.Contains(err.Error(), "400")) {
+		rtspCam, err = NewRTSPCamera(context.Background(), rtspConf, logger)
+	}
+	test.That(t, err, test.ShouldBeNil)
+	stream, err := rtspCam.Stream(context.Background())
+	test.That(t, err, test.ShouldBeNil)
+	_, _, err = stream.Next(context.Background())
+	test.That(t, err, test.ShouldBeNil)
```

Comment:
Why do we do a single call to `Next` here before the loop below?

Reply:
The first call to the camera is different from the rest - it closes a channel called `gotFirstFrame` which releases the process a switch statement it would otherwise get stuck in. I made it separate so that if the first call failed, I would have a better idea of the culprit. I'll add a comment to the test.

Reply:
Oh, gotcha. That makes sense.